pkg/processor/controlcommunication/types.go

```diff
@@ -61,16 +68,50 @@ func (c *ControlConsumer) GetKind() ControlMessageKind {
 	return c.kind
 }
 
-// Send broadcasts a message to all subscribed channels
-func (c *ControlConsumer) Send(message *ControlMessage) error {
+// BroadcastAndCloseSubscriptions sends a message to all subscribed channels and deletes all subscriptions after
+func (c *ControlConsumer) BroadcastAndCloseSubscriptions(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// do nothing if channels is an empty slice
+	if len(c.channels) == 0 {
+		return nil
+	}
+	// write message to the all channels
+	for _, channel := range c.channels {
+		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				if r := recover(); r != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", r)
+				}
+			}()
+			channel <- message
+		}(channel, message)
+	}
+
+	// delete the channel from subscription
+	c.channels = make([]chan *ControlMessage, 0)
+
+	return nil
+}
+
+// Broadcast broadcasts a message to all subscribed channels
+func (c *ControlConsumer) Broadcast(message *ControlMessage) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
 
 	wg := sync.WaitGroup{}
 	wg.Add(len(c.channels))
 	for _, channel := range c.channels {
 
 		go func(channel chan *ControlMessage, message *ControlMessage) {
+			defer func() {
+				if r := recover(); r != nil {
+					fmt.Println("Recovered in BroadcastAndCloseSubscriptions", r)
+				}
+			}()
+			defer wg.Done()
 			channel <- message
-			wg.Done()
 		}(channel, message)
```

Comment:
make this a private function and reuse in both `Broadcast` and `BroadcaseAndCloseSubscription`.

Reply:
@TomerShor Do you mean using a goroutine to make a private function? If so, I don't think it's possible because in one case we use a waitgroup, but in the other case, we don't

Reply:
Right, nvm then